src/main/java/io/confluent/idesidecar/websocket/proxy/ProxyContext.java

```diff
@@ -47,7 +55,51 @@ public ProxyContext withConnection(CCloudConnectionState connection) {
   }
 
   public String getConnectUrl() {
-    // TODO: I guess this won't work for private networks, we'll need something more sophisticated
+    String privateEndpoint = getMatchingPrivateEndpoint();
+    if (privateEndpoint != null) {
+      return transformToLanguageServiceUrl(privateEndpoint);
+    }
+
+    return buildPublicUrl();
+  }
+
+  private String getMatchingPrivateEndpoint() {
+    FlinkPrivateEndpointUtil util = getPrivateEndpointUtil();
+    List<String> endpoints = util.getPrivateEndpoints(environmentId);
+
+    if (endpoints == null || endpoints.isEmpty()) {
+      return null;
+    }
+
+    return endpoints.stream()
+        .filter(endpoint -> util.isValidEndpointWithMatchingRegionAndProvider(endpoint, region, provider))
+        .findFirst()
+        .orElse(null);
+  }
+
+  private FlinkPrivateEndpointUtil getPrivateEndpointUtil() {
+    return CDI.current()
+        .select(FlinkPrivateEndpointUtil.class)
+        .get();
+  }
+
+  /**
+   * Transform a Flink private endpoint URL to a Language Service URL.
+   * Example: https://flink.us-west-2.aws.private.confluent.cloud
+   *       -> wss://flinkpls.us-west-2.aws.private.confluent.cloud/lsp
+   */
+  private String transformToLanguageServiceUrl(String privateEndpoint) {
+    try {
+      java.net.URI uri = java.net.URI.create(privateEndpoint);
```

Comment:
Could we add this as a regular import, `import java.net.URI;`?

```suggestion
      URI uri = URI.create(privateEndpoint);
```

Reply:
Yes, great suggestion! updated.